core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -0,0 +1,291 @@
+    {
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+        this.costCalculator = requireNonNull(costCalculator, "costCalculator is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            TypeProvider types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector,
+            TableStatsProvider tableStatsProvider)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(tableStatsProvider, "tableStatsProvider is null");
+
+        // Skip for write nodes since writing partitioned data with small amount of nodes could cause
+        // memory related issues even when the amount of data is small. Additionally, skip for FTE mode since we
+        // are not using estimated partitionCount in FTE scheduler.
+        if (PlanNodeSearcher.searchFrom(plan).whereIsInstanceOfAny(SKIP_PLAN_NODES).matches()
+                || getRetryPolicy(session) == RetryPolicy.TASK) {
+            return plan;
+        }
+
+        try {
+            return determinePartitionCount(plan, session, types, tableStatsProvider)
+                    .map(partitionCount -> rewriteWith(new Rewriter(partitionCount), plan))
+                    .orElse(plan);
+        }
+        catch (RuntimeException e) {
+            log.warn(e, "Error occurred when determining hash partition count for query %s", session.getQueryId());
+        }
+
+        return plan;
+    }
+
+    private Optional<Integer> determinePartitionCount(PlanNode plan, Session session, TypeProvider types, TableStatsProvider tableStatsProvider)
+    {
+        long minInputRowsPerTask = getMinInputSizePerTask(session).toBytes();
+        if (minInputRowsPerTask == 0) {
+            return Optional.empty();
+        }
+
+        // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+        if (isInputMultiplyingPlanNodePresent(plan)) {
+            return Optional.empty();
+        }
+
+        StatsProvider statsProvider = new CachingStatsProvider(statsCalculator, session, types, tableStatsProvider);
+        CostProvider costProvider = new CachingCostProvider(costCalculator, statsProvider, session, types);
+
+        // Calculate partition count based on nodes output data size
+        double sourceTablesOutputSize = getSourceNodesOutputSize(plan, statsProvider, types);
+        double expandingNodesMaxOutputSize = getExpandingNodesMaxOutputSize(plan, statsProvider, types);
+        if (isNaN(sourceTablesOutputSize) || isNaN(expandingNodesMaxOutputSize)) {
+            return Optional.empty();
+        }
+        int partitionCountBasedOnOutputSize = max(
```

Comment:
Could you do: ```int partitionCountBasedOnOutputSize = getPartitionCountFromOutputSize(max(sourceTablesOutputSize, expandingNodesMaxOutputSize), minInputRowsPerTask)```

Extract `max(sourceTablesOutputSize, expandingNodesMaxOutputSize)` as separate variable